sklearn/decomposition/_fastica.py

```diff
@@ -54,6 +54,10 @@ def _sym_decorrelation(W):
     i.e. W <- (W * W.T) ^{-1/2} * W
     """
     s, u = linalg.eigh(np.dot(W, W.T))
+    # avoid sqrt of negative values or division by zero because of rounding
+    # errors:
+    s = np.clip(s, a_min=np.finfo(W.dtype).eps, a_max=None)
```

Comment:
By extending the existing tests to use the `global_dtype` and the `global_random_seed` fixtures, I found out that this part of the code could break quite easily with negative eigen values.

This fixes it and makes the test much more stable. However I still found fishy things that I marked with `XXX` comments in the diff of this PR.